### ISSUE TITLE
docker ps command hangs

### ISSUE THREAD

**Issue by daniruddha29** [docker/docker#12606]
Hi All,

We have a large setup of postgres databases which are on dockers. The host machine has 3 TB of RAM. When we are trying to execute docker ps command, it hangs and return nothing.
Any help on this is highly appreciable. 

Regards,
Aniruddha

**Comment by crosbymichael** [comment]
i've seen this also

**Comment by daniruddha29** [comment]
any thing you follow, to check what the problem is???

**Comment by LK4D4** [comment]
@daniruddha29 Do you use links or restart policies?

**Comment by daniruddha29** [comment]
no, we don't . Does that impact? 
Hope, this is not serious issue as the environment is production.

**Comment by LK4D4** [comment]
@daniruddha29 Just I recall same issues because of restart and links. I'll try to reproduce, any clues about your containers?

**Comment by daniruddha29** [comment]
we are using RHEL 6.5

**Comment by crosbymichael** [comment]
@daniruddha29 thanks, we can reproduce and will look for a fix and keep you updated

**Comment by LK4D4** [comment]
@daniruddha29 Did you use some volumes for your containers?

**Comment by daniruddha29** [comment]
yes we are using volumes to host our database mount.

**Comment by daniruddha29** [comment]
let me know if you need any specific info.

**Comment by LK4D4** [comment]
@daniruddha29 Any info will be helpful, `docker version`, `docker info`, `Dockerfile`, command line for running containers.

**Comment by daniruddha29** [comment]
docker version 
Client version: 1.2.0
Client API version: 1.14
Go version (client): go1.3.1
Git commit (client): fa7b24f
OS/Arch (client): linux/amd64
Server version: 1.2.0
Server API version: 1.14
Go version (server): go1.3.1
Git commit (server): fa7b24f
++++++++++++++++++++++++++++++++
docker info
Containers: 18
Images: 96
Storage Driver: devicemapper
 Pool Name: docker-252:0-656075-pool
 Pool Blocksize: 64 Kb
 Data file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata file: /var/lib/docker/devicemapper/devicemapper/metadata
 Data Space Used: 18131.5 Mb
 Data Space Total: 102400.0 Mb
 Metadata Space Used: 17.6 Mb
 Metadata Space Total: 2048.0 Mb
Execution Driver: native-0.2
Kernel Version: 3.8.13-44.1.1.el6uek.x86_64
Operating System: <unknown>
+++++++++++++++++++++++++++++++++++++
i will update the docker file shortly

**Comment by daniruddha29** [comment]
its too early to ask.. but did you get anything around that???

**Comment by LK4D4** [comment]
@daniruddha29 Nope, sorry :/ Can you try update your docker?

**Comment by daniruddha29** [comment]
Need to check as this is production environment.
One question, may be very silly one but what will be the impact if continue using it.?

**Comment by LK4D4** [comment]
@daniruddha29 Even if it is unfixed issue it won't be backported to 1.2 :(

**Comment by daniruddha29** [comment]
Haha... definitely not. 
One more, is it a bug ?? already known??

**Comment by LK4D4** [comment]
@daniruddha29 Looks like bug, but we can't reproduce it. Maybe it was fixed since 1.2 version.

**Comment by daniruddha29** [comment]
That is quite useful information.
Thank you very much for all the help.

**Comment by aidanhs** [comment]
Seeing this at the moment:
```
ahobsons@docker03:~$ docker version
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.4.1
Git commit (client): a8a31ef
OS/Arch (client): linux/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): a8a31ef
ahobsons@docker03:~$ docker info
Containers: 24
Images: 298
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 346
Execution Driver: native-0.2
Kernel Version: 3.13.0-46-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 4
Total Memory: 7.63 GiB
Name: docker03
ID: 2JDT:3FKJ:P4FW:C3YE:TQ54:USOD:6X7J:XDF5:SH3A:NTKY:X7GJ:CUUM
WARNING: No swap limit support
```

Machine has spare memory, low load and very little i/o - the daemon looks idle, even when the docker ps command is hung.

Bizarrely, this machine is part of a swarm cluster...and doing docker ps via that works fine.

**Comment by aidanhs** [comment]
Ah, swarm appears to have cached information. New containers I start don't appear.
FWIW, `docker images`, `docker run` work fine when run directly on the machine.

**Comment by LK4D4** [comment]
@aidanhs Yeah, it means that some container acquired lock. So you can't do `ps` because of one locked container, also you can't stop or kill it from cli. Such bugs is very hard to fix without reproduction case :( It is pretty possible that it was fixed in 1.6, because code around running containers was rewritten to use new libcontainer API.

**Comment by aidanhs** [comment]
I assume you're thinking a deadlock then? Or are there other things that can cause this in go? I assume care is taken to always defer unlock after acquiring a lock.

Have you (Docker Inc as a whole, not you specifically) not got any scripts to loop over the goroutines in a running daemon and dump out what they're blocked on? If it was a deadlock that would presumably help.

I've just compiled the daemon with symbols and then loaded those symbols into a gdb attached to the docker daemon. I can now write python scripts to traverse all the goroutines looking for particular things...if I know what I'm looking for. Do you have any links to previous tickets like this?

**Comment by estesp** [comment]
@aidanhs see PR #10786 for dump capability added to Docker daemon.. if you can reproduce easily, add this patch (if possible) and you can then dump goroutines.

**Comment by aidanhs** [comment]
I can dump goroutine traces with gdb. Unfortunately there are 618 goroutines :)
I've scripted their dumping, result at https://gist.github.com/aidanhs/960beaf2db1de622a2dd. This is pretty big.

But! If we grep out the callers to Lock then we find the callsites goroutines are blocked at:
```
root@docker03:~# grep -A1 Lock gdb.txt | grep -v '^--$' | grep -v 'Mutex' | sort | uniq -c
      1 #5  0x00000000004bc5ba in github.com/docker/docker/daemon.(*Daemon).ContainerInspect (daemon=0xc2080691e0, job=0xc208857980, ~r1=0) at /go/src/github.com/docker/docker/daemon/inspect.go:17
      1 #5  0x00000000004c861f in github.com/docker/docker/daemon.(*State).IsRunning (s=0xc2081c79d0, ~r0=54) at /go/src/github.com/docker/docker/daemon/state.go:116
      1 #5  0x00000000004c861f in github.com/docker/docker/daemon.(*State).IsRunning (s=0xc2091c2e70, ~r0=246) at /go/src/github.com/docker/docker/daemon/state.go:116
     41 #5  0x00000000004d4f20 in github.com/docker/docker/daemon.func·025 (container=0xc209060a80, ~r1=...) at /go/src/github.com/docker/docker/daemon/list.go:79
      3 #5  0x00000000004d4f20 in github.com/docker/docker/daemon.func·025 (container=0xc209060fc0, ~r1=...) at /go/src/github.com/docker/docker/daemon/list.go:79
      1 #68 0x00000000005eb5bf in sync.(*Pool).getSlow (p=0x12eac50 <net/http.textprotoReaderPool>, x=...) at /usr/local/go/src/sync/pool.go:129
```

Since I have gdb I can (in theory) print variables from goroutines and so on if you like.
Will have more of a look a bit later. I strongly suspect one of the first three lines is the culprit.

**Comment by LK4D4** [comment]
I personally fail to observe how it can produce a deadlock :)

**Comment by elephantfries** [comment]
@LK4D4  This problem is still present in 1.6.0.  I just had the case yesterday.  We use docker in a build system.  A build hung and the agent abandoned it (without trying to stop or kill the build container). The agent and the build were running in two different containers.  After that, docker ps hung with no way to recover but a reboot.  This was on centos atomic 4a524a58cb with docker 1.6.0.

**Comment by LK4D4** [comment]
@elephantfries It would help if you're described your flags to daemon and `run`.

**Comment by elephantfries** [comment]
Docker deamon (after reboot), from systemctl status docker:

   CGroup: /system.slice/docker.service
           ├─ 843 /usr/bin/docker -d --selinux-enabled --insecure-registry <my.company.registry> --storage-opt dm.loopdatasize=150G
           └─1051 docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 8811 -container-ip 172.17.0.1 -container-port 8811

The agent container, from systemctl status:

   CGroup: /system.slice/agent.service
           └─1009 /usr/bin/docker run -a STDOUT -a STDERR --name agent -p 8811:8811 -h agent23 --security-opt label:disable -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock -v /home/build:/home/build -v /var/srv/data:/opt/data <my.company.registry>/<agent-image> /root/run

The build container, from the build script:

docker run -t --rm --name builder -e JOB=$proj -c 256 -u build -w /home/build --security-opt label:disable -v /opt/data/scripts/docker:/home/build/scripts -v $workdir/build:/home/build/workdir -v /home/build:/home/build/homedir -e MAVEN_OPTS=$maven_opts $(cat builder-image.txt) /home/build/scripts/build-maven $proj $target

No other containers were running.  After the build gave up, a subsequent build started which included a script that kills any left-over containers.  That script uses docker ps to find out what's running.  That script hung and attempts at docker ps from host also hung.

**Comment by elephantfries** [comment]
Sorry, HTML formatting removed some items,  here it is again as intended:


Docker deamon (after reboot), from systemctl status docker:

CGroup: /system.slice/docker.service
├─ 843 /usr/bin/docker -d --selinux-enabled --insecure-registry &lt;my.company.registry&gt; --storage-opt dm.loopdatasize=150G
└─1051 docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 8811 -container-ip 172.17.0.1 -container-port 8811

The agent container, from systemctl status:

CGroup: /system.slice/agent.service
└─1009 /usr/bin/docker run -a STDOUT -a STDERR --name agent -p 8811:8811 -h agent23 --security-opt label:disable -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock -v /home/build:/home/build -v /var/srv/data:/opt/data &lt;my.company.registry/image-name&gt; /root/run

The build container, from the build script:

docker run -t --rm --name builder -e JOB=$proj -c 256 -u build -w /home/build --security-opt label:disable -v /opt/data/scripts/docker:/home/build/scripts -v $workdir/build:/home/build/workdir -v /home/build:/home/build/homedir -e MAVEN_OPTS=$maven_opts $(cat builder-image.txt) /home/build/scripts/build-maven $proj $target

No other containers were running. After the build gave up, a subsequent build started which included a script that kills any left-over containers. That script uses docker ps to find out what's running. That script hung and attempts at docker ps from host also hung.

**Comment by LK4D4** [comment]
So hard to reproduce :( Thanks for info though.
@elephantfries @aidanhs If you would be able to reproduce this at least from time to time. Let me know, we can build debug binary with deadlock detection.

**Comment by elephantfries** [comment]
I do see it from time to time. I should be able to use a special build.

**Comment by aidanhs** [comment]
I've not yet had a chance to use my (still) running process to debug, but I do have some thoughts after a quick look.

In my case, all of the container.Lock operations are blocking because of the two `(*State).IsRunning` stacks - you can see from the full stack dump that these are being called during container start, which locks the container as literally the first thing it does (daemon/container.go `Start`).
So the question now becomes "Why is `state.Lock` deadlocking?".
Scanning daemon/state.go doesn't reveal anything that even calls out to somewhere else, so can't really produce a deadlock. There must be *something* though, as the `state.Lock` lines are definitely hung. I'm not sure how I'd identify where else state gets locked...

**Comment by elephantfries** [comment]
Is the lock held for the entire life of the container?  In my case, the agent is a long running container, it gets locked days after it started. The other containers are either running for a while (30mins),  are in the processes of exiting, or have exited.

**Comment by aidanhs** [comment]
Well, the problem in my case appears to be a deadlock occurring during a container start, and so future attempts to lock that container (ps iterates over all containers, locking them, retrieving information and unlocking them) will also deadlock.

To me it sounds like your build container is the one hanging during startup, which then causes the ps hang? Or are you saying the build container is running a while before it hangs?

**Comment by elephantfries** [comment]
To be precise it is the agent container that hangs.  That container is long running, days.  When it gets a new job it tries to cleanup after old possibly broken builds and runs a script with docker ps to find out if any leftover containers need to be killed before starting any new build containers. In the case described above, there was a previous broken build. Unfortunately, I did not determine what the state of the old build container(s) was.  I suspect that old broken build is critical to creating the deadlock because otherwise the agent is the only container running.

What may also be interesting to note is that the 'broken' build did succeed without trouble next time around so I would say there was something that made the build break and that something may have also deadlocked ps.

I would be happy to run instrumented docker with those builds if available.

**Comment by LK4D4** [comment]
I'll build binary at Monday.

**Comment by aidanhs** [comment]
Got a repro, looking into a fix.

**Comment by LK4D4** [comment]
@elephantfries Do you use `--net=container` in your cases? @aidanhs fixed his case and we're closed issue, but we'll reopent if issue is still present for you.

**Comment by elephantfries** [comment]
No, I don't use --net=container, it seems my case is different, no containers are created at the time of deadlock.  I'd like to keep it open and investigate.

**Comment by LK4D4** [comment]
@elephantfries Here is binary for amd64 with deadlock detection - it'll panic if some mutex is hold too long.
http://s3.dockerproject.com/docker-1.7.0-dev

**Comment by elephantfries** [comment]
@LK4D4 Thanks.  For some reason, it's not starting:

/etc/systemd/system/docker.service.d# systemctl status docker.service 
docker.service - Docker Application Container Engine
   Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled)
  Drop-In: /usr/lib/systemd/system/docker.service.d
           └─flannel.conf
        /etc/systemd/system/docker.service.d
           └─special.conf
   Active: failed (Result: exit-code) since Mon 2015-04-27 19:41:51 UTC; 1min 28s ago
     Docs: http://docs.docker.com
  Process: 5200 ExecStart=/root/docker-1.7.0-dev -d $OPTIONS $DOCKER_STORAGE_OPTIONS $DOCKER_NETWORK_OPTIONS $INSECURE_REGISTRY (code=exited, status=1/FAILURE)
 Main PID: 5200 (code=exited, status=1/FAILURE)

Apr 27 19:41:51 tide.localdomain docker-1.7.0-dev[5200]: time="2015-04-27T19:41:51.705285628Z" level=error msg="Udev sync is not supported. This will lead to ...r-option"
Apr 27 19:41:51 tide.localdomain docker-1.7.0-dev[5200]: time="2015-04-27T19:41:51.705503451Z" level=error msg="[graphdriver] prior storage driver \"devicemap...upported"
Apr 27 19:41:51 tide.localdomain docker-1.7.0-dev[5200]: time="2015-04-27T19:41:51.705953348Z" level=info msg="Listening for HTTP on unix (/var/run/docker.sock)"
Apr 27 19:41:51 tide.localdomain docker-1.7.0-dev[5200]: time="2015-04-27T19:41:51.706590831Z" level=fatal msg="Error starting daemon: error intializing graph...upported"
Apr 27 19:41:51 tide.localdomain systemd[1]: docker.service: main process exited, code=exited, status=1/FAILURE
Apr 27 19:41:51 tide.localdomain systemd[1]: Failed to start Docker Application Container Engine.
Apr 27 19:41:51 tide.localdomain systemd[1]: Unit docker.service entered failed state.
Hint: Some lines were ellipsized, use -l to show in full.

**Comment by LK4D4** [comment]
@elephantfries Oh, you're using devicemapper :( Then you probably need dynamic binary.
Is there overlay fs support in atomic?

**Comment by elephantfries** [comment]
They use devicemapper.  The disk is a straight ext4, no lvm.

**Comment by LK4D4** [comment]
@elephantfries I mean there is probably `overlayfs` support if there is new enough kernel, you can try to run with `-s overlay`.

**Comment by elephantfries** [comment]
Yeah, that works.

**Comment by elephantfries** [comment]
@LK4D4  Looks like I cannot start containers with this image.  From the build log:

14:40:16,799 WARN  - time="2015-04-27T21:40:16Z" level=fatal msg="Error response from daemon: Cannot start container 3164df24e23c8ffbc437568d7df286f98ddc1dceff9f350b73b5f37e0cd13e4f: [8] System error: permission denied"

**Comment by LK4D4** [comment]
@elephantfries Ah, sorry :( I'll see what can I do for building devicemapper version for you. 
Probably you encountered bug with unix-sockets on overlayfs.

**Comment by elephantfries** [comment]
@LK4D4  I see something is crashing my containers. I don't think a bad process would crash a container so I think it is a bug.  This may be unrelated, but I can imagine if it crashes at the right moment, it can potentially leave locks unreleased. It happens sometimes but I have one case where it kills a container every time.  What's the best way to debug that?  I want to make sure it is a bug before opening a ticket.

**Comment by LK4D4** [comment]
I'm not sure tow to debug this without code rebuilding. Lock which is held I think in `daemon/monitor.go`. Maybe you can create some bash script for reproducing, even after many iterations?

**Comment by aidanhs** [comment]
I think it should be possible to use the same method I used?
 - Get a debug binary
 - GDB attach to the deadlocked daemon
 - Load symbols from the debug binary
 - Load go support in GDB
 - Dump a stacktrace of all goroutines
 - Try and reason about which mutex caused the deadlock in the first place

I could note down some more detailed instructions if it'd be helpful.

**Comment by elephantfries** [comment]
Yes, if you don't mind, I am new to go. 
- where do I get debug binary?
- there is no gdb on atomic, how best to get it (yum install not possible)

**Comment by aidanhs** [comment]
I have an e-mail saying there's a response to this issue, but seems to have been deleted?
You should be able to run a container with GDB installed with --privileged and attach to the process that way (I think...it may involve mounting /proc+/sys as well).

I will experiment.

**Comment by elephantfries** [comment]
@aidanhs  Yes, I posted a question which I found an answer to. I thought it was too basic to leave it in the thread.  Anyway, I was able to build a new docker binary and I was able to run gdb and attach to docker. Gdb runs on the host so I don't need --privileged.  The problem is that I cannot 'info goroutines' because runtime-gdb.py fails with errors. There are various fixes filed with golang but each leads to another.  Probably the closest one to a real fix is this:
http://blog.securitymouse.com/2014/10/golang-debugging-turning-pennies-into-gs.html
But even this fails with errors. Looks like RedHat is aware of this (Bug 1192721) but no fix available.  So effectively, I cannot do much with gdb.  I wonder how the docker team managed to work around it.

**Comment by aidanhs** [comment]
Well I made it work by taking runtime-gdb.py from golang master (it's hopelessly broken in the release tarball) and hacking on a tweak to dump all threads at once.

So:
```
# curl https://raw.githubusercontent.com/golang/go/53840ad6/src/runtime/runtime-gdb.py > runtime-gdb.py
[...]
# cp runtime-gdb.py runtime-gdb-hacks.py
# vim runtime-gdb-hacks.py
# ps aux | grep 'docker -d' | grep -v grep
root      1843  1.9  0.4 2460884 39908 ?       Ssl  Apr26 258:18 /usr/bin/docker -d [...]
```

Note that the daemon will be hung whenever gdb is running during the next bits:
```
# gdb --pid=1843 --batch -ex 'maint info sections' | grep '\.text'
 [5]     0x004003e0->0x00ab4984 at 0x000003e0: .text ALLOC LOAD READONLY CODE HAS_CONTENTS
# gdb --pid=1843 --batch -ex 'add-symbol-file docker-1.5.0-debug 0x004003e0' -ex 'source runtime-gdb-hacks.py' -ex 'set logging on' -ex 'info aidan'
```
The final command will output the stacktrace of all goroutines to gdb.txt. As you watch it scroll, you may notice it getting stuck on particular lines - if it seems to have stopped for more than 5s, hit ctrl+c and it'll continue.

You can see the diff that provides 'info aidan' (dumping all goroutines) at https://gist.github.com/aidanhs/a25a6131155834b4c483. Applying the diff should be as simple as `patch runtime-gdb.py gistfile1.diff`.

**Comment by aidanhs** [comment]
(had to tweak my instructions, I suggest you try from scratch if you've tried them before now)

**Comment by elephantfries** [comment]
@aidanhs  Looks like I am hitting the same problem with this version of runtime-gdb.py as well:

gdb -p 31258 --batch -ex 'add-symbol-file docker-1.6.0-wa 0x004003e0' -ex 'source runtime-gdb.py' -ex 'set logging on' -ex 'info aidan'
0x000000000060d504 in syscall.Syscall ()
add symbol table from file "docker-1.6.0-wa" at
	.text_addr = 0x4003e0
Loading Go Runtime support.
Python Exception <class 'gdb.error'> Attempt to extract a component of a value that is not a (null).: 
Error occurred in Python command: Attempt to extract a component of a value that is not a (null).

Your patch applies cleanly.  That's not the problem.  I tried without the patch with the same result:

gdb -p 31258 --batch -ex 'add-symbol-file docker-1.6.0-wa 0x004003e0' -ex 'source runtime-gdb.py' -ex 'set logging on' -ex 'info goroutines'
0x000000000060d504 in syscall.Syscall ()
add symbol table from file "docker-1.6.0-wa" at
	.text_addr = 0x4003e0
Loading Go Runtime support.
Python Exception <class 'gdb.error'> Attempt to extract a component of a value that is not a (null).: 
Error occurred in Python command: Attempt to extract a component of a value that is not a (null).

**Comment by aidanhs** [comment]
Ah ok, docker 1.6.
I've reproduced, will have a look.

**Comment by elephantfries** [comment]
Hit the deadlock right now.  I see (from ps aux) that there are four containers running. This matches what the builder tells me (an agent plus three jobs).  This is, unfortunately, on a centos atomic so I don't have gdb on it.  I could manually copy binary, though. Let me try that.  This is docker 1.6 from centos. 

Any suggestions what to look for?

**Comment by LK4D4** [comment]
@elephantfries I think it is somewhere in daemon/monitor.go. But it is very hard to reproduce. If I'll give you patch with deadlock detection, could you build working binary for atomic?

**Comment by elephantfries** [comment]
Yes, I should be able to build it.

**Comment by elephantfries** [comment]
Looking closer at the build,  there should be another container running: mongodb.  It's not running.  I see mongod daemon is running but there is no corresponding 'docker run'.  This all based on ps aux.  Is it possible a container crashed leaving the deamon running?

**Comment by LK4D4** [comment]
@elephantfries Everything is possible. And this is only way I see to have deadlock now :)
https://github.com/LK4D4/docker/tree/deadlock_detection here branch for detection. But I think that this is case. Maybe you can give us your mongodb image too?

**Comment by elephantfries** [comment]
In that case, we're chasing a crash, not so much a deadlock.   Maybe you can give me a patch to chase a crash?

**Comment by LK4D4** [comment]
@elephantfries Are you running mongodb directly or under some supervisor? I'm not sure how to find such things :/

**Comment by elephantfries** [comment]
/usr/bin/mongod directly

BTW, I just hit another case on another machine.  This time, I was watching it from the beginning and I could see weird things happening long before the deadlock hit.  Specifically: docker ps shows mongo running but ps aux does not.  Then, there is another container 'check-services' which is not showing in docker ps but it is showing in ps aux.  It looks like ps and docker ps got out of sync somehow.

**Comment by LK4D4** [comment]
@elephantfries Do you use some specific flags to run? maybe `--pid=host`?
When there is no container in `docker ps`, but process still alive is bad, but reverse case is much worse. Also, if we'll find deadlock - we can track problems from there.

**Comment by elephantfries** [comment]
All right, I'll try the deadlock patch.

The flags we use:  --name -d -c -e -v -t --rm -u -w --security-opt label:disable

**Comment by elephantfries** [comment]
Hmm.  With the rebuilt docker, which includes the patch, the behavior is different.  As soon as the job is scheduled, the agent container gets killed.  It does not even get a chance to start any containers.  The agent container is gone. It's executable is also gone, nothing in ps aux.

journalctl -u docker from the time a new job was scheduled.

http://pastebin.com/QACV2fhQ

**Comment by elephantfries** [comment]
I think I know why it got killed.  The agent starts all its jobs with an environment variable QBJOBS so that it can clean up by inspecting containers for that env variable. (that's from before labels could be applied)  The agent itself does not define it and there isn't one defined on the host, either.  Now I see that somehow that env variable got propagated to the agent container.  How can this be possible? (also note it was not happening before)

**Comment by elephantfries** [comment]
Never mind the kill.  It was an operator error (wrong image when installing new docker).  It's back to what it was.  Hopefully I get another lockup.

**Comment by aidanhs** [comment]
Ah good, the deadlock branch is based on master so will include #10786. @elephantfries if it deadlocks and it hasn't crashed (i.e. debugmutex has failed to trigger) can you make sure to `kill -USR1` the daemon to get a goroutine trace?
(I can't see how debugmutex would fail, but it doesn't hurt to be prepared)

**Comment by elephantfries** [comment]
@aidanhs Actually, I extracted the patch and applied against centos release, then rebuilt the rpm.  (I want to be as close to the original as possible, there is a chance the culprit may be in RH patches or perhaps device mapper).  I think I'll build another one straight off lk4d4's branch.

**Comment by elephantfries** [comment]
@LK4D4 Looks like your patch detected a 'possible deadlock'.  This time the build just plainly exited because docker quit.  I am assuming this was intentionally triggered by the patch. The log is 614k long.  Here are the last lines:

http://pastebin.com/NNuqquhZ

**Comment by LK4D4** [comment]
@elephantfries Could you provide version-hash, so I can depend on line numbers.

**Comment by LK4D4** [comment]
Pretty weird that there is line about release lock at list.go, but there is no line about acquiring lock at monitor.go :/

**Comment by elephantfries** [comment]
The log is long, 614k, perhaps it was acquired hundreds of lines earlier.

I built it off a centos SRCRPM. It comes with flat extracted files and a reference hash to their repo: 4749651.  The build actually runs git init, to make the sources look like a repo but the hash is meaningless.  I copied the two modified files (the rest are new files, so the lines should match exactly) state.go and vendor.sh:

http://pastebin.com/usgTJ3CP
http://pastebin.com/nsmi4Rhk

**Comment by LK4D4** [comment]
Problem, that it can't be acquired earlier, because in this logs there is acquires and releases.

**Comment by elephantfries** [comment]
I see the last reference to monitor is a request to acquire a lock:

May 18 18:36:08 qba26 docker[1837]: time="2015-05-18T18:36:08Z" level=debug msg="Trying to acquire Lock at /home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/daemon/monitor.go:113"
May 18 18:36:08 qba26 docker[1837]: time="2015-05-18T18:36:08Z" level=debug msg="Lock acquired at /home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/daemon/monitor.go:113"
May 18 18:36:08 qba26 docker[1837]: time="2015-05-18T18:36:08Z" level=debug msg="Stack:\ngoroutine 317 [running]:\ngithub.com/LK4D4/debugmutex.stack(0x0, 0x0)\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/vendor/src/github.com/LK4D4/debugmutex/lock.go:87 +0x85 fp=0xc20889dc60 sp=0xc20889dc10\ngithub.com/LK4D4/debugmutex.(*Mutex).Lock(0xc2084f9700)\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/vendor/src/github.com/LK4D4/debugmutex/lock.go:68 +0x50f fp=0xc20889dde8 sp=0xc20889dc60\ngithub.com/docker/docker/daemon.func·032()\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/daemon/monitor.go:113 +0x6d fp=0xc20889de28 sp=0xc20889dde8\ngithub.com/docker/docker/daemon.(*containerMonitor).Start(0xc20834a000, 0x0, 0x0)\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/daemon/monitor.go:180 +0x7b2 fp=0xc20889df88 sp=0xc20889de28\ngithub.com/docker/docker/daemon.*containerMonitor.Start·fm(0x0, 0x0)\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/daemon/container.go:1419 +0x39 fp=0xc20889dfa8 sp=0xc20889df88\ngithub.com/docker/docker/pkg/promise.func·001()\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/pkg/promise/promise.go:8 +0x2f fp=0xc20889dfe0 sp=0xc20889dfa8\nruntime.goexit()\n\t/usr/lib/golang/src/runtime/asm_amd64.s:2232 +0x1 fp=0xc20889dfe8 sp=0xc20889dfe0\ncreated by github.com/docker/docker/pkg/promise.Go\n\t/home/build/workdir/BUILD/docker-47496519da9664202d900d3635bb840509fa9647/_build/src/github.com/docker/docker/pkg/promise/promise.go:9 +0xfb\n"

**Comment by aidanhs** [comment]
Well it can be acquired earlier because these lock acquisitions could be for different containers - there are no lock memory addresses in the logs (except for the stacktraces) so it's difficult to tell.

I've found one highly suspect piece of code in `daemon/monitor.go`, but I can't 100% see how it causes a deadlock yet. @LK4D4 would it be ok to make a PR without a testcase?

**Comment by elephantfries** [comment]
I thing I see is that those locks seem to be held for a really long time.  I was sitting on a docker ps for over a minute thinking it deadlocked but it came back.  I also see that 'docker ps' exits a lot, lot sooner than its corresponding ps aux 'docker run...'.  The lag can be in minutes.  I measured one to be about six minutes on a task that takes a few seconds.  I know there is '--rm' and the images are large, so it may be legitimate.  Perhaps 5s time out in the detection code is too short and it gives false positives. I think I'll adjust it to 600s.

**Comment by LK4D4** [comment]
@aidanhs Yeah, will be ok.
Long locks bad too, they shouldn't be. But they can be because of disk problems or high system load :(

**Comment by elephantfries** [comment]
This time I have a different failure. It may be related.  It completed but one of my containers did not start because it "Cannot find child".  Here are the lines with the error.  I've got the entire log if needed.

May 19 16:24:25 qba26 docker[7349]: Cannot start container fc02ba2180dde593e7c6e787421269bf5b2b4b98292ca3cc4fcfe40bfe6b8c8f: Cannot find child for /check-services
May 19 16:24:25 qba26 docker[7349]: time="2015-05-19T16:24:25Z" level=info msg="-job start(fc02ba2180dde593e7c6e787421269bf5b2b4b98292ca3cc4fcfe40bfe6b8c8f) = ERR (1)"
May 19 16:24:25 qba26 docker[7349]: time="2015-05-19T16:24:25Z" level=error msg="Handler for POST /containers/{name:.*}/start returned error: Cannot start container fc02ba2180dde593e7c6e787421269bf5b2b4b98292ca3cc4fcfe40bfe6b8c8f: Cannot find child for /check-services"
May 19 16:24:25 qba26 docker[7349]: time="2015-05-19T16:24:25Z" level=error msg="HTTP Error: statusCode=500 Cannot start container fc02ba2180dde593e7c6e787421269bf5b2b4b98292ca3cc4fcfe40bfe6b8c8f: Cannot find child for /check-services"
May 19 16:24:30 qba26 docker[7349]: time="2015-05-19T16:24:30Z" level=error msg="Error scanning log stream: bufio.Scanner: token too long"

**Comment by elephantfries** [comment]
@LK4D4  Well, I've got a real deadlock:

http://pastebin.com/L1PZFgsy

I also think I have a pretty good suspect.  In line 33, devmapper is waiting for 0.  That 0 never arrives to the end of the log and the eventual deadlock.  I see 'waiting for 0' is usually followed by 'destroyed'.  That sequence repeated several times before but not this time. I also see missing 'destroyed' in the case of 'cannot find child' above even though in that case the damage was different.

**Comment by LK4D4** [comment]
@elephantfries Hmm, cool. We can ask devmapper guru @vbatts

**Comment by cgwalters** [comment]
@rhvgoyal also helps maintain devmapper.

**Comment by rhvgoyal** [comment]
@elephantfries 

- Can you provide version information for lvm2, kernel, udev and docker.
- When this hang happens, can you run "dmsetup udevcookies". That will tell us state of existing cookies.
- There is also a way to force completion manually. "dmsetup udevcomplete"

Experimenting with this help narrow down the problem. May be there is an issue with udev here.

**Comment by elephantfries** [comment]
qba26 ~ # rpm -q lvm2 kernel libgudev1
lvm2-2.02.115-3.el7.x86_64
kernel-3.10.0-229.1.2.el7.x86_64
libgudev1-208-20.el7_1.2.x86_64

qba26 ~ # rpm-ostree status
  TIMESTAMP (UTC)         ID             OSNAME                 REFSPEC                                                         
* 2015-04-21 00:08:24     4a524a58cb     centos-atomic-host     centos-atomic-host:centos/7/atomic/x86_64/cloud-docker-host     
  2015-03-10 02:39:02     580f6467b3     centos-atomic-host     centos-atomic-host:centos/7/atomic/x86_64/cloud-docker-host     

qba26 ~ # dmsetup udevcookies
Cookie       Semid      Value      Last semop time           Last change time
0xd4db2de    11730944   1          Tue May 19 17:55:18 2015  Tue May 19 17:55:18 2015

As for docker, it is a custom build off docker-1.6.0-1.el7.src.rpm. Specifically I added LK4D4's patch and rebuilt the rpm.  The base commit is 4749651.

**Comment by rhvgoyal** [comment]
@elephantfries 

So dmsetup udevcookies also shows there is a stuck cookie. The one we are waiting on. Now we need to figure out how did we reach in this situation. 

Do you see anything on console? Wondering if there are any kernel messages there which might give any hint. Otherwise it might be something related to udev rules.

**Comment by rhvgoyal** [comment]
@elephantfries 

This cookie seems to be waiting for creation of device with major 253 and minor 3. Was any such device created at all in /dev/ ?

**Comment by elephantfries** [comment]
There is nothing on the console

dmesg mentions device mapper twice.  This probably matches the two failures: the docker deadlock and the 'cannot find child' mentioned earlier in the thread:

These lines most likely correspond to 'cannot find child':

[84510.711460] docker0: port 1(veth8abe831) entered forwarding state
[84510.712599] docker0: port 1(veth8abe831) entered forwarding state
[84525.728077] docker0: port 1(veth8abe831) entered forwarding state
[84903.401690] device-mapper: ioctl: unable to remove open device docker-8:17-7864323-c1d0d858e3c56701c9b75827702773379c17cb918ec13e93b44c5a1008f651bb-init
[84903.412761] device-mapper: ioctl: unable to remove open device docker-8:17-7864323-c1d0d858e3c56701c9b75827702773379c17cb918ec13e93b44c5a1008f651bb-init
[84912.595270] EXT4-fs (dm-2): mounted filesystem with ordered data mode. Opts: discard
[84914.770410] EXT4-fs (dm-2): mounted filesystem with ordered data mode. Opts: discard
[84915.498681] EXT4-fs (dm-2): mounted filesystem with ordered data mode. Opts: discard
[84915.525218] device veth7123ef8 entered promiscuous mode


These lines most likely correspond to the docker deadlock:


[85653.979474] docker0: port 4(veth064d6bf) entered forwarding state
[85653.980245] docker0: port 4(veth064d6bf) entered forwarding state
[85669.024052] docker0: port 4(veth064d6bf) entered forwarding state
[85717.552484] docker0: port 3(veth0628074) entered disabled state
[85717.557381] device veth0628074 left promiscuous mode
[85717.558914] docker0: port 3(veth0628074) entered disabled state
[85727.546108] device-mapper: ioctl: unable to remove open device docker-8:17-7864323-f187fa761678e19ac1f27830dd78fd8cbff7d2e00414ee96c9604a8b7ca68bf3-init
[85727.556920] device-mapper: ioctl: unable to remove open device docker-8:17-7864323-f187fa761678e19ac1f27830dd78fd8cbff7d2e00414ee96c9604a8b7ca68bf3-init
[85729.187815] EXT4-fs (dm-3): mounted filesystem with ordered data mode. Opts: discard
[85731.091089] EXT4-fs (dm-3): mounted filesystem with ordered data mode. Opts: discard
[85731.433689] EXT4-fs (dm-3): mounted filesystem with ordered data mode. Opts: discard
[85731.462805] device veth17c8048 entered promiscuous mode

**Comment by elephantfries** [comment]
yes:

brw-rw----  1 root disk    253,   3 May 19 17:55 dm-3

**Comment by rhvgoyal** [comment]
@elephantfries 

# stat /dev/dm-3 output please.

**Comment by elephantfries** [comment]
qba26 ~ # stat /dev/dm-3
  File: ‘/dev/dm-3’
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 5h/5d   Inode: 135818      Links: 1     Device type: fd,3
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2015-05-19 17:55:48.503000000 +0000
Modify: 2015-05-19 17:55:18.463000000 +0000
Change: 2015-05-19 17:55:18.463000000 +0000
 Birth: -

**Comment by rhvgoyal** [comment]
Ok, suspicion right now is that there is some udev issue. udev should have called "dmsetup udevcomplete" on this cookie after the creation of node /dev/dm-3. (Look at /usr/lib/udev/rules.d/95-dm-notify.rules).

So now libdm is waiting for udev to decrement reference count (using dmsetup udevcomplete). 

We see that node has been created. But it is not clear that why udevcomplete was not called. Is there a problem with udev rule, or something else.

We probably will have to enable more udev debugging and try to reproduce it and see if we can prove/disprove that udev did call "udevcomplete" or not.

I have not played with it myself, but I have been told that "udevadm control -l debug" can be used to increase verbosity of udev and then re-run the command to create this situation and monitor udev logs and see if something jumps out.

**Comment by elephantfries** [comment]
qba26 ~ # cat /usr/lib/udev/rules.d/95-dm-notify.rules
# Copyright (C) 2009 Red Hat, Inc. All rights reserved.
#
# This file is part of LVM2.

# Udev rules for device-mapper devices.
#
# These rules are responsible for sending a notification to a process
# waiting for completion of udev rules. The process is identified by
# a cookie value sent within "change" and "remove" events (the cookie
# value is set before by that process for every action requested).

ENV{DM_COOKIE}=="?*", RUN+="/usr/sbin/dmsetup udevcomplete $env{DM_COOKIE}"

**Comment by elephantfries** [comment]
@rhvgoyal  Do we still need LK4D4's patch?  If not, I could run original docker.

**Comment by rhvgoyal** [comment]
@elephantfries 

I think running original docker should be sufficient.

**Comment by elephantfries** [comment]
@rhvgoyal Lucky us, I got a failure on the first try.  This is not the deadlock, though, it is the case of 'Cannot find child' but devmapper is in the play gain. This time, it does get udecomplete but there is an operation in the middle of it.  This seems to be enough to cause a failure: killing a container.

In line 22, devmapper is 'waiting for zero' but it does not get it immediately as usual. Instead there is another transaction before 'destroyed' in line 28.  It seems this somehow leads to a failure in line 314.


http://pastebin.com/GLDuSGhN


qba23 ~ # dmsetup udevcookies
Cookie       Semid      Value      Last semop time           Last change time

udevadm monitor appears ordinary, log available if needed.

**Comment by rhvgoyal** [comment]
@elephantfries 

I think this is some other error and not related to device mapper. This message comes from graphdb.go. so there is some inconsistency there. Not sure what operation you are trying to run exactly. If you can break it down into docker commands, might help me make a mental picture better.

If I look at last lines off error, to me it looks like you started container 70c27a8c456c42b78e13c5d91afc860551f617f7218ff6aecb872c9f0b2e6e60 and then stopped it and then when you tried to start it again, you got the error message. 

To me it looks like device mapper finished start operation as well as unmount and device removal operation.

So something else has gone wrong in higher layers. Can't see how this is devicemapper issue.

**Comment by rhvgoyal** [comment]
@elephantfries 

I think we need to narrow it down somehow. Ideally I want to see how to reproduce it myself. If you can come up with some steps on how to reproduce it, that will be helpful.

**Comment by rhvgoyal** [comment]
@elephantfries 

Also can you please tell me about your environement agian. Are you running rhel atomic? Trying to recreate your environment. 

Are you using loop devices or you are using an lvm thin pool. (docker info) will help.

**Comment by elephantfries** [comment]
This is centos atomic:  2015-04-21 00:08:24  4a524a58cb.  No lvm, loop device.

qba23 ~ # docker info
Containers: 1
Images: 49
Storage Driver: devicemapper
 Pool Name: docker-8:17-5242883-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 33.29 GB
 Data Space Total: 161.1 GB
 Data Space Available: 127.8 GB
 Metadata Space Used: 20.18 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.127 GB
 Udev Sync Supported: true
 Data loop file: /var/srv/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/srv/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Kernel Version: 3.10.0-229.1.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 8
Total Memory: 4.844 GiB
Name: qba23
ID: J5TX:6OAP:OJWT:PCXP:5M5T:YVKT:TGBU:QFPZ:4EP2:UXEC:GBOJ:BPDN
Debug mode (server): true
Debug mode (client): false
Fds: 21
Goroutines: 39
System Time: Fri May 22 18:35:40 UTC 2015
EventsListeners: 0
Init SHA1: 90fabd35f7e000f391499e72cb91a6952ace39bf
Init Path: /usr/libexec/docker/dockerinit
Docker Root Dir: /var/srv/docker

**Comment by rhvgoyal** [comment]
Ok. So I am using F21 and running upstream docker on it. One of the things you seem to be doing is "docker start <container_id>".

I enabled debug logs of udev using "udevadm control -l debug". And then I did "docker start <container_id>". I can see following important events in logs.

May 22 14:35:32 tiger docker[23372]: time="2015-05-22T14:35:32.480048996-04:00" level=debug msg="Calling POST /containers/{name:.*}/start"

May 22 14:35:32 tiger docker[23372]: time="2015-05-22T14:35:32.480758573-04:00" level=debug msg="activateDeviceIfNeeded(0303db98519442ac325f68640e9b6952a08ffca84afea4bf53007dbf39cd5f0c)"

May 22 14:35:32 tiger docker[23372]: time="2015-05-22T14:35:32.482071347-04:00" level=debug msg="libdevmapper(7): libdm-common.c:1393 (4) docker-253:0-3014719-0303db98519442ac325f68640e9b6952a08ffca84afea4bf53007dbf39cd5f0c: Stacking NODE_ADD (253,8) 0:6 0660 [verify_udev]"

May 22 14:35:32 tiger systemd-udevd[23696]: starting '/usr/sbin/dmsetup udevcomplete 4234816'

May 22 14:35:32 tiger systemd-udevd[23692]: '/usr/sbin/dmsetup udevcomplete 4234816' [23696] exit with return code 0

May 22 14:35:32 tiger docker[23372]: time="2015-05-22T14:35:32.562660767-04:00" level=debug msg="libdevmapper(7): libdm-common.c:2246 (4) Udev cookie 0xd4d9e40 (semid 28344320) destroyed"

Notice, once I enabled udev debug lots, Now I can see udev is calling "dmsetup udevcomplete" and that leads to destruction of cookie and things move forward.

If you enable udev debugging and try to reproduce and compare success and failure cases, that might give us some clues.

**Comment by rhvgoyal** [comment]
Another thing to try could be switching to lvm thin pool (instead of loop devices) and see if issue is still reproducible. That would tell us whether loop devices are somehow contribute to the problem or not.

**Comment by elephantfries** [comment]
@rhvgoyal  Yeah, I get the 'Udev cookie 0xd4d1a16 (semid 1572864) destroyed' as well, see line 28. I also enabled 'udevadm control -l debug'.  There is log from udevadm monitor but it does not seem interesting.

In successful cases, 'waiting for zero' is immediately followed by 'destroyed'.  For example see lines 8 and 9.

In case of a failure, something gets between in one of the sequences.  See lines 22-28.  In this case 'destroy' does not follow 'wait for 0' immediately.  This is the difference.

**Comment by elephantfries** [comment]
It's very difficult to prove the negative.  For example, I never saw a failure under Vbox.  I saw it only under kvm or bare metal.  Some builds never fail while some seem to be more prone to it.  I was contemplating switching to overlayfs, but doesn't it mask a problem?

**Comment by rhvgoyal** [comment]
If it is a devicemapper problem, yes switching to overlayfs is just masking the issue. 

Now suddenly nature of problem seems to have changed. Initially we were looking at cookie waiting for 0. And looks like that is not reproducible and we are looking at this other error. 

We shall have to reach some sort of stable state where we can narrow it down and reproduce it. As of now, I really can't say where is the problem.

**Comment by elephantfries** [comment]
I can reproduce a deadlock.  It takes longer to, but I am confident I will get another trace.

Is there anything I could do enhance the debugging.

**Comment by rhvgoyal** [comment]
At this point I can only think of enabling udev logs because that seemed to be primary suspect.

**Comment by elephantfries** [comment]
@rhvgoyal Got another deadlock

http://pastebin.com/HtbnZdj2

qba23 ~ # dmsetup udevcookies
Cookie       Semid      Value      Last semop time           Last change time
0xd4d75f2    10715136   1          Fri May 22 20:22:26 2015  Fri May 22 20:22:22 2015


qba23 ~ # stat /dev/dm-5 
  File: ‘/dev/dm-5’
  Size: 0               Blocks: 0          IO Block: 4096   block special file
Device: 5h/5d   Inode: 125768      Links: 1     Device type: fd,5
Access: (0660/brw-rw----)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2015-05-22 20:22:56.570000000 +0000
Modify: 2015-05-22 20:21:57.468000000 +0000
Change: 2015-05-22 20:21:57.468000000 +0000
 Birth: -



Last lines from udevadm monitor related to /dev/dm-5:

KERNEL[11718.866414] add      /devices/virtual/bdi/253:5 (bdi)
KERNEL[11718.867218] add      /devices/virtual/block/dm-5 (block)
UDEV  [11718.869188] add      /devices/virtual/bdi/253:5 (bdi)
UDEV  [11718.869397] add      /devices/virtual/block/dm-5 (block)
KERNEL[11718.882933] change   /devices/virtual/block/dm-5 (block)
UDEV  [11718.900761] change   /devices/virtual/block/dm-5 (block)
KERNEL[11723.997998] change   /devices/virtual/block/dm-5 (block)
KERNEL[11723.998266] remove   /devices/virtual/block/dm-5 (block)
UDEV  [11724.014437] change   /devices/virtual/block/dm-5 (block)
UDEV  [11724.019626] remove   /devices/virtual/block/dm-5 (block)
KERNEL[11724.399871] remove   /devices/virtual/bdi/253:5 (bdi)
KERNEL[11724.400149] remove   /devices/virtual/block/dm-5 (block)
UDEV  [11724.400655] remove   /devices/virtual/bdi/253:5 (bdi)
UDEV  [11724.402841] remove   /devices/virtual/block/dm-5 (block)
KERNEL[11726.572942] add      /devices/virtual/bdi/253:5 (bdi)
KERNEL[11726.573174] add      /devices/virtual/block/dm-5 (block)
UDEV  [11726.573574] add      /devices/virtual/bdi/253:5 (bdi)
UDEV  [11726.577438] add      /devices/virtual/block/dm-5 (block)
KERNEL[11726.592603] change   /devices/virtual/block/dm-5 (block)
UDEV  [11726.620398] change   /devices/virtual/block/dm-5 (block)
KERNEL[11755.686916] change   /devices/virtual/block/dm-5 (block)

**Comment by thaJeztah** [comment]
@LK4D4 @rhvgoyal to get things clear for me; is this a docker issue, a devicemapper or CentOS issue? Just wondering how this should be "labeled" and if this is something for a milestone or not.

**Comment by LK4D4** [comment]
@thaJeztah Can be all three :) If we can fix this - then ok for milestone, if not - then devicemapper is too hard for us :D

**Comment by thaJeztah** [comment]
@LK4D4 thanks, just checking if it would be a blocker for the 1.7 release. If you think it is, please add it to the milestone ;)

**Comment by LK4D4** [comment]
@thaJeztah it existed for whole 1.6 :/ I'm definitely not happy with this, but this isn't blocker.

**Comment by thaJeztah** [comment]
@LK4D4 :+1: thanks, clear

**Comment by rhvgoyal** [comment]
@elephantfries 

Did you enable debug logs for udev using (udevadm control). Do you see any debug messages from udev in journalctl at the time of hang.

**Comment by rhvgoyal** [comment]
@elephantfries 

Also can you test it with latest upstream docker and see if you still face the issue.

**Comment by elephantfries** [comment]
Yes, I did, I listed output from udevadm monitor above.  Journalctl does not show much

qba23 ~ # journalctl -u udevd.sevice
-- Logs begin at Fri 2015-05-22 17:06:31 UTC, end at Tue 2015-05-26 17:01:01 UTC. --

**Comment by rhvgoyal** [comment]
@elephantfries 

Did you enable debug logs from udev using "udevadm control -l debug"? Only then debug logs will show up in journal.

Also I want to know if same issue happens in latest upstream docker or not. May be it is already fixed.

**Comment by elephantfries** [comment]
I can try building from upstream. I'll try that later this evening.  In the meantime, I tried overlayfs.  It works fine with overlayfs but that is not a proof, merely a hint.  Unfortunately, ovelayfs has other bugs which prevent me from using it in production.

**Comment by elephantfries** [comment]
Yes, I did run udevadm control -l debug

**Comment by rhvgoyal** [comment]
@elephantfries 

If you enable debug logs from udev, it does output debug messages. Can you enable logs and try creating a logical volume using "lvcreate" and see if you see udev debug messages or not.

**Comment by rhvgoyal** [comment]
@elephantfries 

Do you have following commit in the sources you are using. I suspect it might have something to do with your issue.

commit 665656afbb8932a11c69c0cd79e21a768aa46d38
Author: Vivek Goyal <vgoyal@redhat.com>
Date:   Thu Apr 2 16:47:14 2015 -0400

    devmapper: Use a pointer as argument to deferred function UdevWait()
    
    UdevWait() is deferred and takes uint cookie as an argument. As arguments
    to deferred functions are calculated at the time of call, it is possible
    that any update to cookie later by libdm are not taken into account when
    UdevWait() is called. Hence use a pointer to uint as argument to UdevWait()
    function.
    
    Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

**Comment by elephantfries** [comment]
I built off docker-1.6.0-1.el7.src.rpm.  It has a tarball with flat file sources with a reference to 4749651. 

Looking at sources, your patch is not there.  For example this is what I see in devmapper.go:

[at line 286]:
func UdevWait(cookie uint) error {
        if res := DmUdevWait(cookie); res != 1 {
                log.Debugf("Failed to wait on udev cookie %d", cookie)
                return ErrUdevWait
        }
        return nil
}

**Comment by elephantfries** [comment]
@rhvgoyal  There is a newer release of centos atomic: 2015-05-15 00:08:39  d7f7a2a2bb.  It's got an updated docker rpm:  docker-1.6.0-11.0.1.el7.centos.x86_64.  It also has a flat file tarball which refers to commit 8aae715d99d7fdeaed1c8043e789d3620520ffef. Looking at sources, this version still does not include your patch.

I could try to rebuild the original docker-1.6.0-1.el7.src.rpm with your patch and see if ti makes a difference. Would I need some other patches for this to work?

**Comment by rhvgoyal** [comment]
no you would not need some other patch for this to work. But we would also need udev debug logs. They appear on my system. You will have to see why udev logs are not there in journal in your system.

**Comment by elephantfries** [comment]
I don't have lvm on this machine.  Any other way to trigger udev debug message?

**Comment by rhvgoyal** [comment]
I think you can try using "dmsetup create"

**Comment by rhvgoyal** [comment]
@elephantfries 

To be specific, udev should be calling something like this on cookie.

May 26 13:41:52 foo systemd-udevd[4689]: starting '/usr/sbin/dmsetup udevcomplete 4246136'
May 26 13:41:52 foo  systemd-udevd[4673]: '/usr/sbin/dmsetup udevcomplete 4246136' [4689] exit with return code 0

**Comment by elephantfries** [comment]
All right.  Mystery solved. The debug messages show with:

journalctl -u systemd-udevd      (not journalctl -u udevd)

Unfortunately, I rebooted the machine so I lost my logs.
Let me try another trace.

**Comment by elephantfries** [comment]
@rhvgoyal  Deadlock 2

qba23 ~ # dmsetup udevcookies
Cookie       Semid      Value      Last semop time           Last change time
0xd4d9b78    4390912    1          Tue May 26 19:51:22 2015  Tue May 26 19:51:22 2015


Docker log last lines:

http://pastebin.com/nSUuczJb

Udevd log last lines:

http://pastebin.com/cqN9DALu

**Comment by rhvgoyal** [comment]
@elephantfries 

I am looking at udev logs and can't find any "dmsetup udevcomplete" calls. In fact, I can't seem to find any messages related to dm in udev logs. Not sure if rules are being executed properly or not.

**Comment by elephantfries** [comment]
Here's the full log, covers the entire build session.

http://pastebin.com/WR1bqQvE

**Comment by elephantfries** [comment]
Sorry, wrng log. This is udevd log:

http://pastebin.com/h2ULXutb

**Comment by rhvgoyal** [comment]
Ok, Looking at udev logs and libdevmapper logs, some observations.

- Last udevcomplete was done on cookie 4248975 (0x40D58F). 
- This seems to match the second last cookie (0xd4dd58f )used for device removal.
- After that docker tried to activate the actual container device and created a new cookie ( 0xd4d9b78) and I don't see a corresponding udevcomplete.

So on the surface it will seem as if udev did not call udevcomplete. Not clear that why did that happen.

**Comment by rhvgoyal** [comment]
@elephantfries 

Can you provide "dmsetup table" output please.

**Comment by elephantfries** [comment]
qba23 ~ # dmsetup table
docker-8:17-5242883-7bed6f54f55c6e5eaeb90dd1282b9e56fa30309e807581495031b423b400d4c3: 0 20971520 thin 253:0 122
docker-8:17-5242883-d0dabf387dc55d67ec3f8ffbbe31bb4ff31b773057876c7b4f635c631cc43bc1: 0 20971520 thin 253:0 111
docker-8:17-5242883-pool: 0 314572800 thin-pool 7:1 7:0 128 32768 1 skip_block_zeroing 
docker-8:17-5242883-3c038f93e80f55e283c9785eb906b768d4c543d51e3ab9d93caffb4bfb7da1a5: 0 20971520 thin 253:0 116
docker-8:17-5242883-d01af288d675afbdea460a55194c7f1fab0426127fd2303c2bbbe7bf965f2a60: 0 20971520 thin 253:0 117
docker-8:17-5242883-7a19daadc8c30d9f311d9e0aa5442b33a22ed3de34d9d9060dde37ec05101cde: 0 20971520 thin 253:0 91

**Comment by elephantfries** [comment]
Switching to using lvm thin pool directly improves the situation.  That solution has its own set of problems but it seems to work for now.  Perhaps the strategy to target overlayfs is best if it is believed to be the successor.

I left a couple of machines in the original configuration in case we want to put some more effort fixing this bug.

**Comment by rhvgoyal** [comment]
@elephantfries 

I had suspicion that udev flags are wrong but that they seem to be right for snapshot device creation.

Alsdair had looked at the logs and noticed that one of the udev thread is being killed and it is possible that probe command failed. Looks like logs are gone now.

using lvm thin pool is the right thing to do. loop devices are unreliable and also docker does not have any management utilities to do pool management (grow pool etc).

BTW, what are the issues you are facing with lvm thin pool?

**Comment by elephantfries** [comment]
@rhvgoyal  I can re-post the logs if it would be useful.

With direct thin pool, it seems that we're limited to 100G.  I tried larger amounts, like 160G, but then I am running into devmapper confusion. I forgot the exact error but somebody else posted it on #atomic list as well.  If you want, I can most likely recreate that problem.  The other issues are minor, manageability type.  I can't complain about them really but to give you an idea what the thinking is for example it forces us to LVM partitions. We'd much rather give 1T to the / partition ext4 and be done with it.  Those atomic hosts are commodities, the simpler the better.  Nobody's expanding disks on them, we'd switch to another host if necessary.  Another example is when we decide to rm -rf /var/lib/docker (because of unrelated issues), then the pool is unusable and we have to re-create it from scratch.  It's the cost of living on the cutting edge of technology, we understand that.

**Comment by elephantfries** [comment]
Here are the logs again:

http://pastebin.com/YKhGV85D
http://pastebin.com/7nrQfysT

**Comment by rhvgoyal** [comment]
@elephantfries 

Ok, you mentioned lot of issues, let us handle the most important one first. That is some kind of this 100G limitation. Looks like I am not facing this issue.

I created an lvm thin pool (using docker-storage-setup) of size 800G and used around 500G for docker thin pool.

I started docker with option --storage-opt dm.basesize=160G and everything seems to be fine. I am
having container root of size 160G.
[root@d2cc496561d6 /]# df -h
Filesystem                                                                                         Size  Used Avail Use% Mounted on

/dev/mapper/docker-253:1-2763198-d2cc496561d6d520cbc0236b4ba88c362c446a7619992123f11c809cded25b47  160G  230M  160G   1% /

I am using upstream docker on F21 and using a dynamically built binary.

So this might be something atomic specific and might have even got fixed into latest. If not, this should be something fixable.

**Comment by rhvgoyal** [comment]
@elephantfries 

The issue of manageability, there are many pieces to it.

- W.r.t planning for dividing space between root and thin pool, I agree that using loop devices is simpler.

- W.r.t being able to grow pool later, I think using loop device is not best as docker does not have ability to grow pool. You mention that you don't care as you will move to a different atomic host. So you don't care about all the containers which are already in the previous thin pool as well?

- Sometimes pool might be in bad shape. lvm checks for the pool health during activation. docker does
   none of it. That means your operations on pool later might fail and you might not have an easy
   or automatic way to deal with it.

- Loop devices are not considered very reliable. Thin pool is more reliable.

IIUC, looks like for you, not having to think about how to partition existing storage is more important than other downsides of loop. I guess its your choice.

**Comment by rhvgoyal** [comment]
@elephantfries 

If you have to do "rm -rf /var/lib/docker", then you have to issue to more commands to delete and recreate pool.
" lvremove docke/docker-pool"
"systemctl start docker-storage-setup"

And that should take care of destroying your existing pool and creating a new one for docker use.

Does not sound like a lot of work?

**Comment by elephantfries** [comment]
@rhvgoyal Right, your points are valid and I said that I cannot really complain about them. I only mention them to give you an idea into what we're thinking.  So, again, nothing wrong with lvm but life is easier without it. Too bad loop devices are unreliable, let's hope overlayfs stabilizes soon.  However, devmapper has other uses, beyond docker.

So let's focus on hard problems.  I'll try to re-create the devmapper confusion with thin pool larger than 100G.  This has been reported and possibly fixed but I'll try anyway.

**Comment by elephantfries** [comment]
@rhvgoyal There may be time to close this issue.  I met vbatts at DockerCon.  He was rather pessimistic about loop devices.  So maybe that's just how it is: won't fix.  I switched to thin pool while setting my sights on overlayfs.  Devmapper with thin pool is not without issues but they're less frequent. I could not reproduce 'no space' on thin pools larger than 100G but I did run into "cannot start container" a couple of times.  Restarting docker brings it back to operation.  Not ideal, but I can live with it.

**Comment by rhvgoyal** [comment]
@elephantfries 

Ok, lets close it. Repoen it again once we have more concrete data we can go after and fix things.

@vbatts, can you please close this.

**Comment by LK4D4** [comment]
Closing

**Comment by syzer** [comment]
on mac i got
```
An error occurred trying to connect: Get https://192.168.59.103:2376/v1.19/containers/json: dial tcp 192.168.59.103:2376: i/o timeout
```

**Comment by oncletom** [comment]
I face a similar issue on an Ubuntu 14.04 EC2 machine – using Docker 1.8.2 with `/var/lib/docker`  mounted on an EBS volume.

I am not sure if it happens during the exited images cleanup (`docker rm $(docker ps -a -q -f "status=exited" -f "status=dead")` every 10 minutes with hundreds of exited containers).

Let me know if you need any additional details.

**Comment by sllawap** [comment]
Also seeing this with CoreOS 766.3.0 and default Docker configuration on Google Compute Engine.

# docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 2c2c52b-dirty
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 2c2c52b-dirty
OS/Arch (server): linux/amd64

# docker info
Containers: 8
Images: 55
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.6-coreos-r1
Operating System: CoreOS 766.3.0
CPUs: 1
Total Memory: 3.618 GiB
Name: <redacted>
ID: ACRN:H7HS:I3O3:NLCJ:22WI:TMHC:GDEX:5RVH:OESO:UCYY:EF53:JHOQ

# ./docker-compose -v
docker-compose version: 1.4.0

I seem to be able to reproduce this by executing many commands of the following form:

# docker-compose run --rm --entrypoint=${command} ${container} /bin/sh -c '<complex command>'

Working on a consistent reproduction case with a simple container.

**Comment by yifan-gu** [comment]
Also seeing this on coreos 766.4

**Comment by thaJeztah** [comment]
@sllawap @yifan-gu the versions you're running on CoreOS are maintained by CoreOS, first of all those are older versions of Docker (which means that the issue might already have been fixed here), but also, the CoreOS may contain patches that make it act different as the "vanilla" Docker version (see the `-dirty` in the version).

I recommend opening an issue in the CoreOS issue tracker first.

**Comment by sjwoodr** [comment]
I also encountered this (again) today with docker 1.8.2 on ubuntu 14.04 in EC2.  I had to reboot.  Restarting docker daemon itself was of no help.

**Comment by dazraf** [comment]
Why is this issue not reopened? Is it really fixed?

**Comment by thaJeztah** [comment]
@dazraf I think this thread has become a "catch all" for multiple issues that may not be related, but have similar results; if you're encountering this on 1.10, it's better to open a new issue

**Comment by alexanderkjeldaas** [comment]
If you're encountering this on 1.10, it might simply be that it is migrating images.

**Comment by kevinoriordan** [comment]
Also seeing this on Amazon Linux 2015.09

Docker version details:

> Client:
>  Version:      1.9.1
>  API version:  1.21
>  Go version:   go1.4.2
>  Git commit:   a34a1d5/1.9.1
>  Built:        
>  OS/Arch:      linux/amd64
> 
> Server:
>  Version:      1.9.1
>  API version:  1.21
>  Go version:   go1.4.2
>  Git commit:   a34a1d5/1.9.1
>  Built:        
>  OS/Arch:      linux/amd64

**Comment by jhberges** [comment]
Also seeing this on 

    docker@xxxx:~$ uname -a
    Linux xxxx 3.16.0-45-generic #60~14.04.1-Ubuntu SMP Fri Jul 24 21:16:23 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

    docker@xxxx:~$ docker version
    Client:
     Version:      1.10.3
     API version:  1.22
     Go version:   go1.5.3
     Git commit:   20f81dd
     Built:        Thu Mar 10 15:54:52 2016
     OS/Arch:      linux/amd64
    
    Server:
     Version:      1.10.3
     API version:  1.22
     Go version:   go1.5.3
     Git commit:   20f81dd
     Built:        Thu Mar 10 15:54:52 2016
     OS/Arch:      linux/amd64

**Comment by brouberol** [comment]
Also seeing this on 

```
# docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:38:58 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:38:58 2016
 OS/Arch:      linux/amd64
```

The funny thing is, I have NO running container. I initially thought this hang could be related to a high number of containers running.

**Comment by LouisKottmann** [comment]
Also seeing this on:

```
# docker version
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:30:23 2016
 OS/Arch:      linux/amd64
# uname -a
Linux ip-10-0-1-66 3.13.0-86-generic #130-Ubuntu SMP Mon Apr 18 18:27:15 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.4 LTS
Release:        14.04
Codename:       trusty
```

**Comment by thaJeztah** [comment]
@brouberol @LouisKottmann please open a new issue, and also provide the output of `docker info`.


I'm going to lock this issue; this issue has become a collection of many issue that are not related to each other, but can result in the same behavior ("docker not responding"). If you're having an issue like this, please open a new issue, and try to provide information that can be used to reproduce, or debug. Reporting only "I have this too" likely won't be sufficient to find the cause, or to see if there's an actual *bug* or if it's a configuration issue.
